tests: metrics: formalise the metrics reporting methods

grahamwhaley opened the pull request:
The metrics report through more than one method at present (csv files, stdout). Add a more formal and flexible method for reporting metrics results so we can then migrate all the metrics over to 'one true method'.

Fixes: #954

grahamwhaley commented:
@MarioCarrilloA Please take a look as well.
@GabyCT You may be interested, as I intend to move all the network metrics to use this interface as one of the next actions.

chavafg commented:
qa-passed

chavafg commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+	fi
+
+	if [ -z "$IMG" ];then
+		IMG="$(readlink "$cc_img_path/clear-containers.img")"
```
we can use `@CONTAINERS_IMG@` to prevent hardcoding the path to `/usr/share/clear-containers`

grahamwhaley replied:
Right. Was aware of the hardcoding. I'll look at moving to a `.in` file. I may pick up #945 before I do that then to make development easier (I was hoping to delay that for a bit, oh well ;-)

chavafg commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+	fi
+
+	if [ -z "$KERNEL" ];then
+		KERNEL="$(readlink "$cc_img_path/vmlinux.container")"
```
same as above, we can use `@CONTAINER_KERNEL@`

MarioCarrilloA replied:
Hi @chavafg , if we use @CONTAINER_KERNEL@ we are forced to build the project in order to expand that variable right?, I think the PnP tools should run independently the project is build or not, what do you think?

grahamwhaley replied:
I think that is true @MarioCarrilloA , and I am coding a bit of 'backup' into the script right now - but, the ultimate code would locate exactly what runtime the test ran with, and then extract the info about the img file and kernel from that runtime setup. Sadly, we are some way from coding that up today - so, today we will go with @xxx@, and 'adapt'. It may be for now that the person or system who runs the tests may have to set some ENV or command line vars to point to the correct img and kernel files.

chavafg replied:
@MarioCarrilloA yes, you would need to build the project. For usability, I would say that you are right, you shouldn't need build the whole project for just running the metrics. 
But right now other scripts needed here depend from building the project, so you will still need to run `make metrics-tests`.

grahamwhaley replied:
See #945 the intention being that a simple make target (intermediate step) be introduced so we can just generate the pre-processed files without having to build the whole project.

jodh-intel commented on tests/metrics/README.md:
```diff
@@ -12,19 +20,60 @@ To run the metrics, please use the Makefile rule `metrics-tests` on the top leve
+| Result   | The result of the test (without unit tags)         |
+| Units    | The Units the Result is measured in (KB, S etc.)   |
+
+The back end of the `save_results` function is to calls into the `send_results.sh` script found in the
```
s/is to calls/calls/

jodh-intel commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+function main()
+{
+	local OPTIND
+	while getopts ":hdvn:r:s:o:a:u:g:t:w:i:k:" opt;do
```
I'd sort this list alphabetically to make it easier to maintain.

jodh-intel commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+		   help
+		   exit 0;
+		   ;;
+		n)
```
I'd also sort these case statements for easier maintenance.

jodh-intel commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+	fi
+
+	if [ -z "$RESULT" ];then
+		echo "error: result no found"
```
s/no found/not found/.

jodh-intel commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+	fi
+
+	if [ -z "$UNITS" ];then
+		echo "error: units no found"
```
s/no found/not found/.

jodh-intel commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+
+	if [ -z "$TAG" ];then
+		if [ -f "$runtime" ];then
+			commit="$(cc-oci-runtime -v | tail -n1 | sed s/"commit: "//g)"
```
Do we envisage this script even moving to the tests repo? If so, it would be worth creating a variable for the runtime (so we could switch it to `cc-runtime` and this test would then become:

```
$runtime -v|grep "commit.*:"|cut -d: -f2-|tr -d ' '
```
(works for `cc-oci-runtime` and `cc-runtime`).

jodh-intel commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+	CSV_FILE=${RESULT_DIR}/$(echo ${TEST_NAME} | sed 's/ /-/g').csv
+
+	# Are we dry-running?
+	if [ -z "$SEND" ];then
```
I like the idea of having a dry-run mode, but I'm not really clear what it buys us here. I think it would just perform the checks? It might be worth having it "almost" run the code below by showing `would run '...'` using an `echo $cmd` and if not in dry-run mode, running `eval $cmd` instead.

jodh-intel commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+			echo "Timestamp,Group,Name,Args,Result,Units,System,SystemVersion,Platform,Image,Kernel,Commit" > ${CSV_FILE}
+		fi
+
+		s1=$(echo "$timestamp,$GROUP,$TEST_NAME,$ARGS,$RESULT,$UNITS,$SYSTEM,$SYS_VERSION,$platform,$IMG,$KERNEL,$commit")
```
Strictly, the code should handle commas in these variables - either error out or quote the values. A simple check might be to `echo $1` and count the commas to ensure we have the expected number of fields.

jodh-intel commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+	 This tool will save results to csv
+	 files.
+   Options:
+	 -h   Help page
```
It would be good if this list of options was sorted alphabetically :smile:

jodh-intel commented:
Oh - could you add that `Fixes: #954` into one of the commits? That will soon be required (once #952 lands).

grahamwhaley commented:
@jodh-intel thanks for the thorough review - the script was pulled in from another project. I did some mods and cleanups, but looks like there is a bunch more to do to meet our requirements :-) Let me get on that and send a new version soon.
Yeah, I mean to add the 'fixes'... thanks for picking that up as well.

grahamwhaley commented:
Jodh - oh, iswym with wanting the fixes in the commit, not the merge/PR message. Hmm, I can - but, personally it feels a bit 'wrong', on a couple of fronts:
- In a multi-commit series, the Fix is really spread across all of them, so you would have to put the fixes comment on the last commit. It feels 'better' in/on the PR/merge message to me
- I don't particularly like putting github specific info into the code commits themselves - that sort of ties our git repo directly to github - which, sure, we may never have an intention of moving anywhere else - but, it just feels wrong to me.

Just noting like. I'll follow suit and add the fixes at the next update.

jodh-intel commented:
@grahamwhaley - I can understand that, but -- I think -- for most the workflow is easier if you do have a "Fixes:" comment in one of the commits since you don't need to adjust the PR text and the wokflow becomes minimally:

- push branch
- click to raise a PR
- have github auto-close the issue for you when the PR is merged.

I take the point about a stack of commits, but I'd tend to mark those that contribute to fixing an issue as "partially fixes #XXX". Again, I understand not wanting to encode github specifics in the branch, but as shown above, it *is* useful to use a "Fixes:" comment to make life easier for developers. I think it's not unreasonable to have close ties between the version control system and change control systems. The final point is that if we only put the fixes on the PR itself, `checkcommits` can't perform the fixes check so wouldn't add value there.

Flipping it round, I'd say we want to minimise the amount of extra detail we add to a PR description since that isn't stored as history in the git branch.

chavafg commented on tests/lib/send_results.sh:
```diff
@@ -0,0 +1,202 @@
+function save_to_csv()
+{
+	runtime="/usr/bin/cc-oci-runtime"
+	release_file="/usr/lib/os-release"
```
This path works forfor Ubuntu, Fedora and Clear Linux. Seems that CentOS does not have it.

jodh-intel replied:
... and of course if we only check `/etc/`, it won't work for clr ;-( It's almost like this needs to be a `getent` database... ;-) So maybe special-case CentOS (as it's the oldest)?

dlespiau replied:
https://www.freedesktop.org/software/systemd/man/os-release.html

grahamwhaley commented on tests/metrics/README.md:
````diff
@@ -12,19 +20,60 @@ To run the metrics, please use the Makefile rule `metrics-tests` on the top leve
+```bash
+$ make metrics-tests
+```
+Note, this command will also try to execute all the tests - you may wish to halt execution once the
````
Correct

grahamwhaley commented on tests/metrics/README.md:
```diff
@@ -12,19 +20,60 @@ To run the metrics, please use the Makefile rule `metrics-tests` on the top leve
+| Name     | The name of the test. This name is generally used to form a unique test identifer if the data is ultimately stored |
+| Args     | Any arugments that were passed to the test         |
+| Result   | The result of the test (without unit tags)         |
+| Units    | The Units the Result is measured in (KB, S etc.)   |
```
Hi. I suspect we have some confusion here - let's check. the save_results function takes four arguments. One of them (the second one) is a list of 'arguments' that were used for that particular test run and need storing with the result data (iyswim). I suspect having 'Args' as an argument is confusing here. I suggest I could either:
 - rename 'Args' to 'TestParms' or similar to avoid the confusion
 - "double quote" the arguments on the command description like this: `save_results "Name" "Args" "Result" "Units"` to clarify that they are four individual arguments.
Does that make sense, and will either of those help clarify @jodh-intel

jodh-intel replied:
Oh - I think we might have been talking at crossed purposes - all I was saying was that the description for "Units" for example mentions "Result" which I took to be a reference to the "Result" column. If so, we should be consistent and make all those description lines use a leading upper-case character for any of the arguments in the table. For example, "Result" would become:

```
| Result   | The Result of the test (without unit tags)         |
```

But I'm happy with your plan on the rename + quoting.

grahamwhaley replied:
OK, I've done some mods - let's see how the new one 'feels.
Still some confusion though - to note, there is no 'Result column' (did you mean 'row'?), and using 'Result' in the description for the Result row is somewhat self referential - it is a description of the content, 'result' is correct. I did change 'Units' to 'units' in the below Unit row though - as that one was wrong...

grahamwhaley commented:
@jodh-intel @chavafg all feedback should have been addressed in the new version.

grahamwhaley commented:
and then I spot I've acquired a Makefile.am conflict... so, expect a rebase and another push...

chavafg commented:
qa-passed

chavafg commented:
qa-passed

iphutch commented:
Looks good from a docs perspective. Awaiting Makefile.am conflict resolution to approve.

jodh-intel commented on tests/lib/send_results.sh.in:
```diff
@@ -0,0 +1,339 @@
+		# This is somewhat imperfect. Ideally we'd have knowledge passed
+		# in from the test itself about which runtime it used so we could
+		# be certain to have the correct runtime and extract the correct
+		# commit id
```
Worth capturing in an issue?

grahamwhaley replied:
Indeed. Added #973

jodh-intel commented:
lgtm

grahamwhaley commented:
@chavafg any chance you can re-review so we can move this on? thanks.

chavafg commented on tests/lib/test-common.bash.in:
```diff
@@ -63,6 +64,22 @@ function write_result_to_file(){
+		die "save_results() requires 4 parameters, not $#"
+	fi
+
+	$LIB_DIR/send_results.sh -n "$1" -a "$2" -r "$3" -u "$4"
```
I tested the PR and when calling this line I got: 

/root/go/src/github.com/01org/cc-oci-runtime/tests/metrics/density/../../lib/test-common.bash: line 80: /root/go/src/github.com/01org/cc-oci-runtime/tests/metrics/density/../../lib/send_results.sh: Permission denied

It can be fixed with: 

```
bash $LIB_DIR/send_results.sh -n "$1" -a "$2" -r "$3" -u "$4"
```

grahamwhaley replied:
ah, indeed, good spot. A side effect of the .in file processing. My local version got chmod'd during test, and then that 'stuck'. thx. will fix and update.

grahamwhaley commented:
Updated @chavafg @jodh-intel.
 - added the `bash` invocation of the sub script
 - Also added translation of '/' chars to '-'s in the result file name generation (I realised we are quite likely to hit this when some tests have arguments such as '/bin/true'
 - Fixed a single trailing whitespace typo in the README.

jodh-intel commented:
lgtm

chavafg commented:
qa-passed

chavafg commented:
lgtm
